README.md
# Smooth Scroll behavior polyfill

The [Scroll Behavior specification](https://developer.mozilla.org/en/docs/Web/CSS/scroll-behavior) has been introduced as an extension of the `Window` interface to allow for the developer to opt in to native smooth scrolling. To date this has only been implemented in _Firefox_.

Check out all the methods covered here: https://iamdustan.github.io/smoothscroll


## Installation and use

Download the production ready file [here](https://unpkg.com/smoothscroll-polyfill/dist/smoothscroll.min.js) and include it in your project, or install it as a package.

```sh
# npm
npm install smoothscroll-polyfill --save

# yarn
yarn add smoothscroll-polyfill
```

When including the polyfill in a script tag, it will run immediately after loaded. If you are importing it as a dependency, make sure to call the `polyfill` method:

```js
import smoothscroll from 'smoothscroll-polyfill';

// kick off the polyfill!
smoothscroll.polyfill();
```

In both cases, the script will detect if the spec is natively supported and take action only when necessary.

_The code requires requestAnimationFrame polyfill for browsers which don't support it._


### Force polyfill implementation

If you prefer to always override the native scrolling methods, place this global variable before requiring the module or including the polyfill file.

```js
window.__forceSmoothScrollPolyfill__ = true;
```

_We strongly recommend not to do this unless your project strongly needs it._


## Contribute

The requirements to contribute are [yarn](https://yarnpkg.com) and the latest LTS [Node.js](https://nodejs.org/en/) version.

First, fork the repository and do `yarn install` in the root folder to get all the dependencies to work with. Create a feature branch, write your stuff and run `yarn test` to check code style and prevent bugs.

In this project we use [Prettier](https://prettier.io) to format the final published code, you can run `yarn format` before committing. If you don't, a precommit hook will prevent you from pushing code that hasn't been formatted properly.

Are you done? Awesome, submit a pull request explaining your changes.

_This is a polyfill, not library, so make sure the behavior you are introducing is in the spec._

On tests files we are using ES2015, but the polyfill is written in ES5 for browser compatibility.


### Watch tests

If you want to watch tests as you write your code run `yarn test --watch`.


## Browser Support

Successfully working in:

- _natively supported in Chrome_
- _natively supported in Firefox_
- Safari 6+
- Internet Explorer 9+
- Microsoft Edge 12+
- Opera Next


## Standards documentation

- http://dev.w3.org/csswg/cssom-view
- http://lists.w3.org/Archives/Public/www-style/2013Mar/0314.html


CHANGELOG.md

# <= 0.4.3

* Support `scrollIntoView` on `document.body`
* Support scroll behaviors in shadow DOM

# <= 0.4.2

* ¯\_(ツ)_/¯


test/support/raf.js

// http://paulirish.com/2011/requestanimationframe-for-smart-animating/
// http://my.opera.com/emoller/blog/2011/12/20/requestanimationframe-for-smart-er-animating

// requestAnimationFrame polyfill by Erik Möller. fixes from Paul Irish and Tino Zijdel

// MIT license

(function() {
var lastTime = 0;
var vendors = ['ms', 'moz', 'webkit', 'o'];
for(var x = 0; x < vendors.length && !window.requestAnimationFrame; ++x) {
    window.requestAnimationFrame = window[vendors[x]+'RequestAnimationFrame'];
    window.cancelAnimationFrame = window[vendors[x]+'CancelAnimationFrame']
                               || window[vendors[x]+'CancelRequestAnimationFrame'];
}

if (!window.requestAnimationFrame)
    window.requestAnimationFrame = function(callback, element) {
        var currTime = new Date().getTime();
        var timeToCall = Math.max(0, 16 - (currTime - lastTime));
        var id = window.setTimeout(function() { callback(currTime + timeToCall); },
          timeToCall);
        lastTime = currTime + timeToCall;
        return id;
    };

if (!window.cancelAnimationFrame)
    window.cancelAnimationFrame = function(id) {
        clearTimeout(id);
    };
}());


test/smoothscroll.test.js
import clone from 'lodash.clonedeep';
import smoothscroll from '../src/smoothscroll';
import test from 'ava';

// mock window object
const mockedWindow = {
  Element: {
    prototype: {
      scroll: f => f,
      scrollBy: f => f,
      scrollIntoView: f => f
    }
  },
  scroll: f => f,
  scrollBy: f => f,
  navigator: {
    userAgent: 'test'
  },
  performance: {
    now: f => f
  },
  requestAnimationFrame: f => f
};

// mock document object
const mockedDocument = {
  documentElement: {
    style: {}
  }
};

// expose clean browser objects as global
test.beforeEach(() => {
  global.window = clone(mockedWindow);
  global.document = clone(mockedDocument);
});

// clean browser objects after each test
test.afterEach.always(() => {
  global.window = null;
  global.document = null;
});

test('polyfill is available as method', t => {
  t.truthy(typeof smoothscroll.polyfill === 'function');
});

test('polyfill overrides native methods when polyfill is not supported', t => {
  const originalWindowScroll = window.scroll;
  const originalWindowScrollBy = window.scrollBy;
  const originalElScroll = window.Element.prototype.scroll;
  const originalElScrollBy = window.Element.prototype.scrollBy;
  const originalElScrollIntoView = window.Element.prototype.scrollIntoView;

  smoothscroll.polyfill();

  // global methods were replaced
  t.not(window.scroll, originalWindowScroll);
  t.not(window.scrollBy, originalWindowScrollBy);
  t.not(window.Element.prototype.scroll, originalElScroll);
  t.not(window.Element.prototype.scrollBy, originalElScrollBy);
  t.not(window.Element.prototype.scrollIntoView, originalElScrollIntoView);
});

test('polyfill bails out when polyfill is supported', t => {
  const originalWindowScroll = window.scroll;
  const originalWindowScrollBy = window.scrollBy;
  const originalElScroll = window.Element.prototype.scroll;
  const originalElScrollBy = window.Element.prototype.scrollBy;
  const originalElScrollIntoView = window.Element.prototype.scrollIntoView;

  // global methods remained untouched
  document.documentElement.style.scrollBehavior = 'auto';

  smoothscroll.polyfill();

  t.is(window.scroll, originalWindowScroll);
  t.is(window.scrollBy, originalWindowScrollBy);
  t.is(window.Element.prototype.scroll, originalElScroll);
  t.is(window.Element.prototype.scrollBy, originalElScrollBy);
  t.is(window.Element.prototype.scrollIntoView, originalElScrollIntoView);
});

test('polyfill overrides native methods when implementation is forced', t => {
  const originalWindowScroll = window.scroll;
  const originalWindowScrollBy = window.scrollBy;
  const originalElScroll = window.Element.prototype.scroll;
  const originalElScrollBy = window.Element.prototype.scrollBy;
  const originalElScrollIntoView = window.Element.prototype.scrollIntoView;

  // simulate smooth scroll spec native support
  document.documentElement.style.scrollBehavior = 'auto';

  // force implementation
  window.__forceSmoothScrollPolyfill__ = true;

  smoothscroll.polyfill();

  // global methods were forcefully replaced
  t.not(window.scroll, originalWindowScroll);
  t.not(window.scrollBy, originalWindowScrollBy);
  t.not(window.Element.prototype.scroll, originalElScroll);
  t.not(window.Element.prototype.scrollBy, originalElScrollBy);
  t.not(window.Element.prototype.scrollIntoView, originalElScrollIntoView);
});


test/support/test-core.js
function $(selector) {
    return document.querySelector(selector);
}

function later(cb) {
  setTimeout(cb, 600)
}

function fail(msg) {
    document.querySelector('.test-result').textContent += msg
}

function done(msg) {
    var resultCtr = document.querySelector('.test-result')
    if (resultCtr.textContent === "") {
        resultCtr.classList.add('success')
    }
    resultCtr.textContent += "complete"
}

function assertScroll(actual, expected) {
    // Allow a little bit of leeway
    if (Math.abs(actual - expected) > 3) {
        fail(":failure: Expected '" + actual + "' to be '" + expected + "':")
    }
}

function parseQuery(qstr) {
    var query = {};
    var a = qstr.substr(1).split('&');
    for (var i = 0; i < a.length; i++) {
        var b = a[i].split('=');
        query[decodeURIComponent(b[0])] = decodeURIComponent(b[1] || '');
    }
    return query;
}


gulpfile.js
'use strict';

var gulp = require('gulp'),
    $ = require('gulp-load-plugins')(),
    project = require('./package.json');

var paths = {
  src: './src/' + project.title + '.js',
  output: './dist'
};

var banner = '/*' +
  '\n * ' + project.title + ' polyfill - v' + project.version +
  '\n * ' + project.homepage +
  '\n * ' + project.copyright + ' (c) ' + project.author.name + ', ' + project.contributors[0].name +
    ' - ' + project.license + ' License' +
  '\n */\n\n';

// lint task
gulp.task('lint', function() {
  return gulp.src(paths.src)
    .pipe($.eslint())
    // outputs the lint results to the console.
    .pipe($.eslint.format())
    // To have the process exit with an error code (1) on
    // lint error, return the stream and pipe to failOnError last.
    .pipe($.eslint.failOnError());
});

// build task
gulp.task('build', [ 'lint' ], function() {
  return gulp.src([ paths.src ])
    .pipe($.concatUtil.header(banner))
    .pipe(gulp.dest(paths.output));
});

// assign default gulp task
gulp.task('default', [ 'build' ]);


Gruntfile.js
'use strict'

module.exports = function(grunt) {
  // load all grunt tasks
  require('matchdep').filterDev('grunt-*').forEach(grunt.loadNpmTasks)

  grunt.initConfig({
    jshint: {
      options: {
        jshintrc: '.jshintrc'
      },
      all: [
        'smoothscroll.js'
      ]
    },
    uglify: {
      core: {
        files: {
          'dist/smoothscroll.js': ['smoothscroll.js']
        }
      },
      gzip: {
        files: {
          'dist/smoothscroll.js': ['smoothscroll.js']
        },
        options: {
          report: 'gzip'
        }
      }
    },
    concat: {
      options: {
        separator: ';',
      },
      dist: {
        src: ['bower_components/raf.js/raf.min.js', 'dist/smoothscroll.js'],
        dest: 'dist/smoothscroll.raf.js',
      }
    }
  });

  grunt.registerTask('default', [
    'jshint'
  ]);

  grunt.registerTask('build', [
    'jshint',
    'uglify:core',
    'concat'
  ]);

  grunt.registerTask('build:report', [
    'jshint',
    'uglify:gzip',
    'concat'
  ]);
}




smoothScrollTo.js
(function () {
  'use strict';

  if ('scrollBehavior' in document.documentElement.style) return;

  var originalScrollTo = window.scrollTo;
  var originalScrollIntoView = Element.prototype.scrollIntoView;

  function now() {
    return window.performance !== undefined && window.performance.now !== undefined ? window.performance.now() : Date.now !== undefined ? Date.now() : +new Date;
  }

  // ease-in-out
  function ease(k) {
    return 0.5 * (1 - Math.cos(Math.PI * k));
  }

  var startY, startX, endX, endY;
  window.scrollTo = function(x, y, type) {
    if (type !== 'smooth')
      return originalScrollTo(x, y);

    // TODO: make this intelligent. 300ms per scroll
    var SCROLL_TIME = 300;
    var frame;
    var sx = window.pageXOffset;
    var sy = window.pageYOffset;

    if (typeof startX === 'undefined') {
      startX = sx;
      startY = sy;
      endX = x;
      endY = y;
    }

    var startTime = now();

    // TODO: look into polyfilling scroll-behavior: smooth css property
    // var ey = element.offsetTop, ex = element.offsetLeft;
    var step = function() {
      var time = now();
      var elapsed = (time - startTime) / SCROLL_TIME;
      elapsed = elapsed > 1 ? 1 : elapsed;

      var value = ease(elapsed);
      var cx = sx + ( x - sx ) * value;
      var cy = sy + ( y - sy ) * value;

      originalScrollTo(cx, cy)

      if (cx === endX && cy === endY) {
        startX = startY = endX = endY = undefined;
        return;
      }
      frame = requestAnimationFrame(step);
    }

    frame = requestAnimationFrame(step);
  }

  Element.prototype.scrollIntoView = function(toTop, behavior) {
    if (behavior !== 'smooth') return originalScrollIntoView(toTop, behavior);
    if (typeof toTop === 'undefined') toTop = true;


    if (toTop)
      return window.scrollTo(this.offsetLeft, this.offsetTop, behavior);

    return window.scrollTo(this.offsetLeft, this.offsetTop - document.documentElement.clientHeight + this.clientHeight, behavior)
  }

}());



src/smoothscroll.js
'use strict';

// polyfill
function polyfill() {
  // aliases
  var w = window;
  var d = document;

  // return if scroll behavior is supported and polyfill is not forced
  if (
    'scrollBehavior' in d.documentElement.style &&
    w.__forceSmoothScrollPolyfill__ !== true
  ) {
    return;
  }

  // globals
  var Element = w.HTMLElement || w.Element;
  var SCROLL_TIME = 468;

  // object gathering original scroll methods
  var original = {
    scroll: w.scroll || w.scrollTo,
    scrollBy: w.scrollBy,
    elementScroll: Element.prototype.scroll || scrollElement,
    scrollIntoView: Element.prototype.scrollIntoView
  };

  // define timing method
  var now =
    w.performance && w.performance.now
      ? w.performance.now.bind(w.performance)
      : Date.now;

  /**
   * indicates if a the current browser is made by Microsoft
   * @method isMicrosoftBrowser
   * @param {String} userAgent
   * @returns {Boolean}
   */
  function isMicrosoftBrowser(userAgent) {
    var userAgentPatterns = ['MSIE ', 'Trident/', 'Edge/'];

    return new RegExp(userAgentPatterns.join('|')).test(userAgent);
  }

  /*
   * IE has rounding bug rounding down clientHeight and clientWidth and
   * rounding up scrollHeight and scrollWidth causing false positives
   * on hasScrollableSpace
   */
  var ROUNDING_TOLERANCE = isMicrosoftBrowser(w.navigator.userAgent) ? 1 : 0;

  /**
   * changes scroll position inside an element
   * @method scrollElement
   * @param {Number} x
   * @param {Number} y
   * @returns {undefined}
   */
  function scrollElement(x, y) {
    this.scrollLeft = x;
    this.scrollTop = y;
  }

  /**
   * returns result of applying ease math function to a number
   * @method ease
   * @param {Number} k
   * @returns {Number}
   */
  function ease(k) {
    return 0.5 * (1 - Math.cos(Math.PI * k));
  }

  /**
   * indicates if a smooth behavior should be applied
   * @method shouldBailOut
   * @param {Number|Object} firstArg
   * @returns {Boolean}
   */
  function shouldBailOut(firstArg) {
    if (
      firstArg === null ||
      typeof firstArg !== 'object' ||
      firstArg.behavior === undefined ||
      firstArg.behavior === 'auto' ||
      firstArg.behavior === 'instant'
    ) {
      // first argument is not an object/null
      // or behavior is auto, instant or undefined
      return true;
    }

    if (typeof firstArg === 'object' && firstArg.behavior === 'smooth') {
      // first argument is an object and behavior is smooth
      return false;
    }

    // throw error when behavior is not supported
    throw new TypeError(
      'behavior member of ScrollOptions ' +
        firstArg.behavior +
        ' is not a valid value for enumeration ScrollBehavior.'
    );
  }

  /**
   * indicates if an element has scrollable space in the provided axis
   * @method hasScrollableSpace
   * @param {Node} el
   * @param {String} axis
   * @returns {Boolean}
   */
  function hasScrollableSpace(el, axis) {
    if (axis === 'Y') {
      return el.clientHeight + ROUNDING_TOLERANCE < el.scrollHeight;
    }

    if (axis === 'X') {
      return el.clientWidth + ROUNDING_TOLERANCE < el.scrollWidth;
    }
  }

  /**
   * indicates if an element has a scrollable overflow property in the axis
   * @method canOverflow
   * @param {Node} el
   * @param {String} axis
   * @returns {Boolean}
   */
  function canOverflow(el, axis) {
    var overflowValue = w.getComputedStyle(el, null)['overflow' + axis];

    return overflowValue === 'auto' || overflowValue === 'scroll';
  }

  /**
   * indicates if an element can be scrolled in either axis
   * @method isScrollable
   * @param {Node} el
   * @param {String} axis
   * @returns {Boolean}
   */
  function isScrollable(el) {
    var isScrollableY = hasScrollableSpace(el, 'Y') && canOverflow(el, 'Y');
    var isScrollableX = hasScrollableSpace(el, 'X') && canOverflow(el, 'X');

    return isScrollableY || isScrollableX;
  }

  /**
   * finds scrollable parent of an element
   * @method findScrollableParent
   * @param {Node} el
   * @returns {Node} el
   */
  function findScrollableParent(el) {
    while (el !== d.body && isScrollable(el) === false) {
      el = el.parentNode || el.host;
    }

    return el;
  }

  /**
   * self invoked function that, given a context, steps through scrolling
   * @method step
   * @param {Object} context
   * @returns {undefined}
   */
  function step(context) {
    var time = now();
    var value;
    var currentX;
    var currentY;
    var elapsed = (time - context.startTime) / SCROLL_TIME;

    // avoid elapsed times higher than one
    elapsed = elapsed > 1 ? 1 : elapsed;

    // apply easing to elapsed time
    value = ease(elapsed);

    currentX = context.startX + (context.x - context.startX) * value;
    currentY = context.startY + (context.y - context.startY) * value;

    context.method.call(context.scrollable, currentX, currentY);

    // scroll more if we have not reached our destination
    if (currentX !== context.x || currentY !== context.y) {
      w.requestAnimationFrame(step.bind(w, context));
    }
  }

  /**
   * scrolls window or element with a smooth behavior
   * @method smoothScroll
   * @param {Object|Node} el
   * @param {Number} x
   * @param {Number} y
   * @returns {undefined}
   */
  function smoothScroll(el, x, y) {
    var scrollable;
    var startX;
    var startY;
    var method;
    var startTime = now();

    // define scroll context
    if (el === d.body) {
      scrollable = w;
      startX = w.scrollX || w.pageXOffset;
      startY = w.scrollY || w.pageYOffset;
      method = original.scroll;
    } else {
      scrollable = el;
      startX = el.scrollLeft;
      startY = el.scrollTop;
      method = scrollElement;
    }

    // scroll looping over a frame
    step({
      scrollable: scrollable,
      method: method,
      startTime: startTime,
      startX: startX,
      startY: startY,
      x: x,
      y: y
    });
  }

  // ORIGINAL METHODS OVERRIDES
  // w.scroll and w.scrollTo
  w.scroll = w.scrollTo = function() {
    // avoid action when no arguments are passed
    if (arguments[0] === undefined) {
      return;
    }

    // avoid smooth behavior if not required
    if (shouldBailOut(arguments[0]) === true) {
      original.scroll.call(
        w,
        arguments[0].left !== undefined
          ? arguments[0].left
          : typeof arguments[0] !== 'object'
            ? arguments[0]
            : w.scrollX || w.pageXOffset,
        // use top prop, second argument if present or fallback to scrollY
        arguments[0].top !== undefined
          ? arguments[0].top
          : arguments[1] !== undefined
            ? arguments[1]
            : w.scrollY || w.pageYOffset
      );

      return;
    }

    // LET THE SMOOTHNESS BEGIN!
    smoothScroll.call(
      w,
      d.body,
      arguments[0].left !== undefined
        ? ~~arguments[0].left
        : w.scrollX || w.pageXOffset,
      arguments[0].top !== undefined
        ? ~~arguments[0].top
        : w.scrollY || w.pageYOffset
    );
  };

  // w.scrollBy
  w.scrollBy = function() {
    // avoid action when no arguments are passed
    if (arguments[0] === undefined) {
      return;
    }

    // avoid smooth behavior if not required
    if (shouldBailOut(arguments[0])) {
      original.scrollBy.call(
        w,
        arguments[0].left !== undefined
          ? arguments[0].left
          : typeof arguments[0] !== 'object' ? arguments[0] : 0,
        arguments[0].top !== undefined
          ? arguments[0].top
          : arguments[1] !== undefined ? arguments[1] : 0
      );

      return;
    }

    // LET THE SMOOTHNESS BEGIN!
    smoothScroll.call(
      w,
      d.body,
      ~~arguments[0].left + (w.scrollX || w.pageXOffset),
      ~~arguments[0].top + (w.scrollY || w.pageYOffset)
    );
  };

  // Element.prototype.scroll and Element.prototype.scrollTo
  Element.prototype.scroll = Element.prototype.scrollTo = function() {
    // avoid action when no arguments are passed
    if (arguments[0] === undefined) {
      return;
    }

    // avoid smooth behavior if not required
    if (shouldBailOut(arguments[0]) === true) {
      // if one number is passed, throw error to match Firefox implementation
      if (typeof arguments[0] === 'number' && arguments[1] === undefined) {
        throw new SyntaxError('Value could not be converted');
      }

      original.elementScroll.call(
        this,
        // use left prop, first number argument or fallback to scrollLeft
        arguments[0].left !== undefined
          ? ~~arguments[0].left
          : typeof arguments[0] !== 'object' ? ~~arguments[0] : this.scrollLeft,
        // use top prop, second argument or fallback to scrollTop
        arguments[0].top !== undefined
          ? ~~arguments[0].top
          : arguments[1] !== undefined ? ~~arguments[1] : this.scrollTop
      );

      return;
    }

    var left = arguments[0].left;
    var top = arguments[0].top;

    // LET THE SMOOTHNESS BEGIN!
    smoothScroll.call(
      this,
      this,
      typeof left === 'undefined' ? this.scrollLeft : ~~left,
      typeof top === 'undefined' ? this.scrollTop : ~~top
    );
  };

  // Element.prototype.scrollBy
  Element.prototype.scrollBy = function() {
    // avoid action when no arguments are passed
    if (arguments[0] === undefined) {
      return;
    }

    // avoid smooth behavior if not required
    if (shouldBailOut(arguments[0]) === true) {
      original.elementScroll.call(
        this,
        arguments[0].left !== undefined
          ? ~~arguments[0].left + this.scrollLeft
          : ~~arguments[0] + this.scrollLeft,
        arguments[0].top !== undefined
          ? ~~arguments[0].top + this.scrollTop
          : ~~arguments[1] + this.scrollTop
      );

      return;
    }

    this.scroll({
      left: ~~arguments[0].left + this.scrollLeft,
      top: ~~arguments[0].top + this.scrollTop,
      behavior: arguments[0].behavior
    });
  };

  // Element.prototype.scrollIntoView
  Element.prototype.scrollIntoView = function() {
    // avoid smooth behavior if not required
    if (shouldBailOut(arguments[0]) === true) {
      original.scrollIntoView.call(
        this,
        arguments[0] === undefined ? true : arguments[0]
      );

      return;
    }

    // LET THE SMOOTHNESS BEGIN!
    var scrollableParent = findScrollableParent(this);
    var parentRects = scrollableParent.getBoundingClientRect();
    var clientRects = this.getBoundingClientRect();

    if (scrollableParent !== d.body) {
      // reveal element inside parent
      smoothScroll.call(
        this,
        scrollableParent,
        scrollableParent.scrollLeft + clientRects.left - parentRects.left,
        scrollableParent.scrollTop + clientRects.top - parentRects.top
      );

      // reveal parent in viewport unless is fixed
      if (w.getComputedStyle(scrollableParent).position !== 'fixed') {
        w.scrollBy({
          left: parentRects.left,
          top: parentRects.top,
          behavior: 'smooth'
        });
      }
    } else {
      // reveal element in viewport
      w.scrollBy({
        left: clientRects.left,
        top: clientRects.top,
        behavior: 'smooth'
      });
    }
  };
}

if (typeof exports === 'object' && typeof module !== 'undefined') {
  // commonjs
  module.exports = { polyfill: polyfill };
} else {
  // global
  polyfill();
}


smoothscroll.raf.js
!function(e){for(var n=0,t=["webkit","moz"],a=e.requestAnimationFrame,i=e.cancelRequestAnimationFrame,m=t.length;--m>=0&&!a;)a=e[t[m]+"RequestAnimationFrame"],i=e[t[m]+"CancelRequestAnimationFrame"];a&&i||(a=function(e){var t=+Date.now(),a=Math.max(n+16,t);return setTimeout(function(){e(n=a)},a-t)},i=clearTimeout),e.requestAnimationFrame=a,e.cancelRequestAnimationFrame=i}(window);

(function () {
  'use strict';

  if ('scrollBehavior' in document.documentElement.style) return;

  var originalScrollTo = window.scrollTo;
  var originalScrollIntoView = Element.prototype.scrollIntoView;

  function now() {
    return window.performance !== undefined && window.performance.now !== undefined ? window.performance.now() : Date.now !== undefined ? Date.now() : new Date().getTime();
  }

  // ease-in-out
  function ease(k) {
    return 0.5 * (1 - Math.cos(Math.PI * k));
  }

  var startY, startX, endX, endY;
  window.scrollTo = function(x, y, type) {
    if (type !== 'smooth')
      return originalScrollTo(x, y);

    // TODO: make this intelligent based on distance. 300ms per scro
    var SCROLL_TIME = 300;
    var frame;
    var sx = window.pageXOffset;
    var sy = window.pageYOffset;

    if (typeof startX === 'undefined') {
      startX = sx;
      startY = sy;
      endX = x;
      endY = y;
    }

    var startTime = now();

    // TODO: look into polyfilling scroll-behavior: smooth css property
    // var ey = element.offsetTop, ex = element.offsetLeft;
    var step = function() {
      var time = now();
      var elapsed = (time - startTime) / SCROLL_TIME;
      elapsed = elapsed > 1 ? 1 : elapsed;

      var value = ease(elapsed);
      var cx = sx + ( x - sx ) * value;
      var cy = sy + ( y - sy ) * value;

      originalScrollTo(cx, cy);

      if (cx === endX && cy === endY) {
        startX = startY = endX = endY = undefined;
        return;
      }

      frame = requestAnimationFrame(step);
    };

    frame = requestAnimationFrame(step);
  };

  Element.prototype.scrollIntoView = function(toTop, behavior) {
    if (behavior !== 'smooth')
      return originalScrollIntoView(toTop, behavior);

    if (typeof toTop === 'undefined')
      toTop = true;


    if (toTop)
      return window.scrollTo(this.offsetLeft, this.offsetTop, behavior);

    return window.scrollTo(
      this.offsetLeft,
      this.offsetTop - document.documentElement.clientHeight + this.clientHeight,
      behavior
    );
  };

}());

